app/services/duplicate_profile_resolver.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+# A service class to merge duplicate Profile objects
+class DuplicateProfileResolver
+  class << self
+    def run!
+      @profiles = nil
+      duplicate_profiles.find_each do |profile|
+        if existing_profile(profile)
+          migrate_profile(existing_profile(profile), profile)
+        else
+          self.existing_profile = profile
+        end
+      end
+    end
+
+    private
+
+    def existing_profile(profile)
+      profiles[[profile.ref_id, profile.account_id, profile.benchmark_id]]
+    end
+
+    def existing_profile=(profile)
+      profiles[[profile.ref_id,
+                profile.account_id,
+                profile.benchmark_id]] = profile
+    end
+
+    def profiles
+      @profiles ||= {}
+    end
+
+    def migrate_profile(existing_p, duplicate_p)
+      migrate_profile_hosts(existing_p, duplicate_p)
+      migrate_test_results(existing_p, duplicate_p)
+      migrate_child_profiles(existing_p, duplicate_p)
+      duplicate_p.destroy # BusinessObjectives, ProfileRules
+    end
+
+    def duplicate_profiles
+      Profile.joins(
+        "JOIN (#{grouped_nonunique_profile_tuples.to_sql}) as p on "\
+        'profiles.ref_id = p.ref_id AND '\
+        'profiles.account_id is not distinct from p.account_id AND '\
+        'profiles.benchmark_id = p.benchmark_id'
+      )
+    end
+
+    def grouped_nonunique_profile_tuples
+      Profile.select(:ref_id, :account_id, :benchmark_id)
+             .group(:ref_id, :account_id, :benchmark_id)
+             .having('COUNT(id) > 1')
+    end
```

Comment:
Currently from prod:
```
irb(main):017:0> duplicate_profiles.to_sql
=> "SELECT \"profiles\".* FROM \"profiles\" JOIN (SELECT \"profiles\".\"ref_id\", \"profiles\".\"account_id\", \"profiles\".\"benchmark_id\" FROM \"profiles\" GROUP BY \"profiles\".\"ref_id\", \"profiles\".\"account_id\", \"profiles\".\"benchmark_id\" HAVING (COUNT(id) > 1)) as p on profiles.ref_id = p.ref_id AND profiles.account_id is not distinct from p.account_id AND profiles.benchmark_id = p.benchmark_id"
irb(main):018:0> duplicate_profiles
=> #<ActiveRecord::Relation []>
```

Reply:
I suspect this won't be the case after running the duplicate benchmark resolver, but my latest PR with tweaks to that one should be a bit smarter about creating duplicates: https://github.com/RedHatInsights/compliance-backend/pull/397

Reply:
With the updated benchmark uniqueness PR, https://github.com/RedHatInsights/compliance-backend/pull/397, this should continue to be true, hopefully resulting in very fast migrations. :100: